crates/sui-adapter/src/adapter.rs

```diff
@@ -497,25 +502,38 @@ fn process_successful_execution<
                     module_id,
                     &mut object_owner_map,
                     &newly_generated_ids,
+                    &mut newly_generated_unused,
+                    &mut frozen_object_ids,
+                    &mut child_count_deltas,
                 )
             }
             EventType::DeleteObjectID => {
                 // unwrap safe because this event can only be emitted from processing
                 // native call delete_id, which guarantees the type of the id.
                 let uid: UID = bcs::from_bytes(&event_bytes).unwrap();
                 let obj_id = uid.object_id();
+                newly_generated_unused.remove(obj_id);
                 // We don't care about IDs that are generated in this same transaction
                 // but only to be deleted.
                 if !newly_generated_ids.contains(obj_id) {
                     match by_value_objects.remove(obj_id) {
-                        Some((_, version)) => {
+                        Some((owner, version, child_count)) => {
                             state_view.log_event(Event::delete_object(
                                 module_id.address(),
                                 module_id.name(),
                                 ctx.sender(),
                                 *obj_id,
                             ));
-                            state_view.delete_object(obj_id, version, DeleteKind::Normal)
+                            if let Owner::ObjectOwner(parent_id) = owner {
```

Comment:
Why don't we check here whether the object to be deleted has 0 child at this point? Wonder if that would simplify things (always check child count at deletion point)

Reply:
We might not have processed the children yet. It is so you can delete all children and all parents in the same txn.

We could instead process the transfers children-first, but it would require some sorting here

Reply:
Shouldn't the events be processed in the order they are created?

Reply:
Do we want to enforce that the parents must be deleted exactly before the children? It feels like an unnecessary footgun. Especially when a lot of this complexity is still needed for dealing with wrapped objects.

Reply:
Did you mean "the parents must be deleted after all the children were deleted"?

Reply:
yes, sorry

Reply:
That seems a reasonable requirement?

Reply:
First, it doesn't make the code any easier. As we only know if an object has been wrapped at the end. So this code won't get get any less complex.

Second, I don't think it is a great requirement, because there is nothing in the Move layer that indicates you will bump into this. We have the ability to easily make people's lives easier without changing the complexity of the code, so why not?